Persist ColorPicker color mode

aole opened the pull request:
Color mode will be saved and retrieved from project metadata in EditorSettings.
Need for using a default color picker has been eliminated as this will set the last used mode by the user.

Fixes #30755 and #30754

![test](https://user-images.githubusercontent.com/4707543/63981620-d688cd80-ca8d-11e9-8da6-14be5948c256.gif)

aole commented:
@Calinou @akien-mga let me know if this is reasonable solution.

aole commented:
> scene/gui/color_picker.cpp:330:22: error: use of undeclared identifier
>      'EditorSettings'
>        String color_mode = EditorSettings::get_singleton()->get_project...

Not sure why the checks are throwing this error.

Calinou commented:
@aole This solution sounds good to me :slightly_smiling_face:

YeldhamDev commented:
That's because `EditorSettings` doesn't exist outside the editor. Use `#ifdef TOOLS_ENABLED`.

aole commented:
@Calinou @YeldhamDev Thanks!

I have blocked out all the code for TOOLS_ENABLED. This would not be useful for export templates anyways.

akien-mga commented:
See #29730, we shouldn't have editor code at all in `scene/`.

IMO it's better to just add an editor setting that controls the default mode for all ColorPicker instances in the editor (as suggested in #30755), and use that in the editor when instantiating ColorPickers, without having to change ColorPicker code.

aole commented:
@akien-mga Something like this?

![image](https://user-images.githubusercontent.com/4707543/64031138-d16f6100-cb15-11e9-9950-d6606cdba2f9.png)

Where should it appear? after Custom Display Scale or some place else?

akien-mga commented:
> Where should it appear? after Custom Display Scale or some place else?

I'd put it in the Editor > Inspector section.

>  But I am not sure how to invoke the color picker for rendering/environment/default_clear_color in the project editor. As it seems to be set in main.cpp.

I guess the only code that should have to be changed is for `EditorPropertyColor` in `editor/editor_properties.cpp`.

aole committed:
Default ColorPicker color mode setting

There is a new setting for Default ColorPicker color mode in Editor Settings->Interface->Inspector. Initially this setting will be RGB. Editor ColorPicker will always start with mode defined in this settiing.
Fixes #30755 and #30754

merumelu commented:
There are two other ColorPickers in the editor where this should be added:
https://github.com/godotengine/godot/blob/4f6eb3610f70b2a73a19dd8b92fb03c4739ef25b/editor/property_editor.cpp#L857
https://github.com/godotengine/godot/blob/4f6eb3610f70b2a73a19dd8b92fb03c4739ef25b/editor/plugins/script_text_editor.cpp#L1739

aole commented:
@merumelu I'll add those too.

aole commented:
This is the functionality now ...

![pull31782](https://user-images.githubusercontent.com/4707543/64050883-5b83ed80-cb47-11e9-9ac1-33649b96b55b.gif)

@akien-mga Thanks for the hint.
Let me know if this is what  you were expecting.

merumelu commented on editor/editor_properties.cpp:
```diff
@@ -1864,6 +1864,13 @@ EditorPropertyColor::EditorPropertyColor() {
+	// get default color picker mode from editor settings
+	int default_color_mode = EDITOR_GET("interface/inspector/default_color_picker_mode");
+	if (default_color_mode == 1)
+		picker->get_picker()->set_hsv_mode(true);
```
I wonder if this has any performance impacts. ColorPickerButton doesn't create the picker until it is needed but calling `get_picker()` will initialize one all the time.
https://github.com/godotengine/godot/blob/4f6eb3610f70b2a73a19dd8b92fb03c4739ef25b/scene/gui/color_picker.cpp#L989-L994
The comment even mentions that it was done this way because of the property inspector.

aole replied:
I could do it in 

> ColorPickerButton->get_picker()

 but that would mean going outside the editor which was the case in the earlier commit.

merumelu replied:
Maybe ColorPickerButton should emit a signal once the ColorPicker is ready? Then you can connect to that.

aole replied:
@merumelu That is indeed a better solution.
Made changes accordingly. Please review.

akien-mga commented:
Thanks!
